Objects/longobject.c

```diff
@@ -4511,16 +4511,17 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
         if (newsize <= 0)
             return PyLong_FromLong(0);
         hishift = PyLong_SHIFT - remshift;
-        lomask = ((digit)1 << hishift) - 1;
-        himask = PyLong_MASK ^ lomask;
         z = _PyLong_New(newsize);
         if (z == NULL)
             return NULL;
-        for (i = 0, j = wordshift; i < newsize; i++, j++) {
-            z->ob_digit[i] = (a->ob_digit[j] >> remshift) & lomask;
-            if (i+1 < newsize)
-                z->ob_digit[i] |= (a->ob_digit[j+1] << hishift) & himask;
+        j = wordshift;
+        accum = a->ob_digit[j++] >> remshift;
+        for (i = 0; j < Py_SIZE(a); i++, j++) {
+            accum |= (twodigits)a->ob_digit[j] << hishift;
+            z->ob_digit[i] = (digit)(accum & PyLong_MASK);
```

Comment:
Would not be better to operate on a single digit? `(digit)accum & PyLong_MASK`

Reply:
Maybe. I find this code clearer as a statement of intent: we're only changing the value once, not twice (all other things being equal, I like my casts not to change values).

I'll run some timings and look at the generated code. If the cast-first variant is faster, I'll change this.